### PR TITLE
Test fixes for semantic_text

### PR DESCRIPTION
**Pull request by ioanatia**
Because `semantic_text` is available only in snapshots some of the tests fail when running with `-Dbuild.snapshot=false` since I did not add explicit checks for `EsqlCapabilities.Cap.SEMANTIC_TEXT`.
I assumed that adding `semantic_text` in `UNDER_CONSTRUCTION` would take care of that.

Examples of release test failing on a different PR: https://github.com/elastic/elasticsearch/pull/114831

To check I have run:

```
./gradlew ":x-pack:plugin:esql:test" -Dbuild.snapshot=false  -Dbuild.snapshot=false -Dlicense.key=x-pack/license-tools/src/test/resources/public.key
```

### CODE REVIEW

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/convert/AbstractConvertFunction.java [hunk 1]
```diff
@@ -78,7 +79,15 @@ protected TypeResolution resolveType() {
     }
 
     public Set<DataType> supportedTypes() {
-        return factories().keySet();
+        Set<DataType> supportedTypes = new HashSet<>();
```
Comment (ioanatia) [inline x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/convert/AbstractConvertFunction.java:82]: it's either this check here or adding a capabilities check for each function/operator on the evaluators list.
if that's the case we should probably do the same for `date_nanos` as well.
but I preferred this check than modifying the evaluators list.

**Comment by ioanatia** [comment]
currently checking a much simpler fix by just removing this check:

https://github.com/elastic/elasticsearch/blob/409fb8db7235be96c0cfdf8cf3514ddfdf0aa8ba/x-pack/plugin/esql-core/src/main/java/org/elasticsearch/xpack/esql/core/type/DataType.java#L372-L377

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/convert/AbstractConvertFunction.java [hunk 2]
```diff
@@ -78,7 +79,15 @@ protected TypeResolution resolveType() {
...
+        for (DataType dataType : factories().keySet()) {
+            // TODO: this should use DataType.UNDER_CONSTRUCTION but we would need to fix the failures for date_nanos
+            if (dataType == DataType.SEMANTIC_TEXT && EsqlCapabilities.Cap.SEMANTIC_TEXT_TYPE.isEnabled() == false) {
+                continue;
+            }
+            supportedTypes.add(dataType);
+        }
+        return supportedTypes;
```
Comment (craigtaverner) [inline x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/convert/AbstractConvertFunction.java:90]: Would this be more concise, and perhaps easier to read?

```
        return EsqlCapabilities.Cap.SEMANTIC_TEXT_TYPE.isEnabled()
            ? factories().keySet()
            : factories().keySet().stream().filter(dataType -> dataType != DataType.SEMANTIC_TEXT).collect(Collectors.toSet());
```

**Review by craigtaverner — APPROVED** [review]
I expected this change to mostly impact unit tests, which it seems to have done. I did have a suggestion for the `supportedTypes` method, which you could consider.

**Comment by ioanatia** [comment]
closing in favour of https://github.com/elastic/elasticsearch/pull/116202

**Pull request closed without merging**